### PR TITLE
Client: Add SRV-based virtual hosting support

### PR DESCRIPTION
**Pull request by hoffie**
**Short description of changes**
- Internal: Refactor NetworkUtil::ParseNetworkAddress
  This moves the host elaborate and reusable host name lookup logic to its own ParseNetworkAddressName method.

- Client: Add SRV-based virtual hosting support
  When connecting to host names without port (e.g. example.org), the client will now perform an SRV lookup to _jamulus._udp.example.org. If this lookup returns exactly one result, the result is used to select the actual target address/port.

  If the lookup does not return a usable result, the regular connect logic kicks in (A lookup, default port).

CHANGELOG: Client: Add SRV-based virtual hosting support

**Context: Fixes an issue?**
Related: https://github.com/orgs/jamulussoftware/discussions/1772

**Does this change need documentation? What needs to be documented and how?**
Website, Server administration

**Status of this Pull Request**
Ready for testing/discussion.

**What is missing until this pull request can be merged?**
Ready for testing/discussion.

## Checklist

-  [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
-  [x] I tested my code and it does what I want
-  [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
-  [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
-  [x] I've filled all the content above

### CODE REVIEW

File: src/util.cpp [hunk 1]
```diff
@@ -889,6 +889,97 @@ QSize CMinimumStackedLayout::sizeHint() const
...
+    {
+        HostAddress = CHostAddress ( InetAddr, record.port() );
+        return true;
+    }
+    return false;
+}
+
+bool NetworkUtil::ParseNetworkAddressWithSrvDiscovery ( QString strAddress, CHostAddress& HostAddress, bool bEnableIPv6 )
+{
+    // Try SRV-based discovery first:
```
Comment (hoffie) [inline src/util.cpp:974]: Trying SRV first is the saner behavior. It comes at the cost that everyone has to do the SRV lookup and wait for the (possibly empty) reply.

Comment (pljones) [reply src/util.cpp:974]: Agreed.  The DNS server should fail fairly fast if the entry isn't there.  At worst it should double the lookup time.

File: src/util.cpp [hunk 2]
```diff
@@ -889,6 +889,97 @@ QSize CMinimumStackedLayout::sizeHint() const
 * Other Classes                                                                *
 \******************************************************************************/
 // Network utility functions ---------------------------------------------------
+bool NetworkUtil::ParseNetworkAddressName ( QString strAddress, QHostAddress& InetAddr, bool bEnableIPv6 )
+{
+    // try to get host by name, assuming
+    // that the string contains a valid host name string
```
Comment (pljones) [inline src/util.cpp:895]:
```suggestion
    // that the string contains a valid host name string or IP address
```

File: src/util.cpp [hunk 3]
```diff
@@ -889,6 +889,97 @@ QSize CMinimumStackedLayout::sizeHint() const
 * Other Classes                                                                *
 \******************************************************************************/
 // Network utility functions ---------------------------------------------------
+bool NetworkUtil::ParseNetworkAddressName ( QString strAddress, QHostAddress& InetAddr, bool bEnableIPv6 )
```
Comment (pljones) [inline src/util.cpp:892]: I'm not keen on this method name.  It doesn't explain that it's the one place we turn any string into an IP address.  I think I'd prefer `ParseNetworkAddressString`.

File: src/util.cpp [hunk 4]
```diff
@@ -889,6 +889,97 @@ QSize CMinimumStackedLayout::sizeHint() const
...
+{
+    // Try SRV-based discovery first:
+    if ( ParseNetworkAddressSrv ( strAddress, HostAddress, bEnableIPv6 ) )
+    {
+        return true;
+    }
+    // Try regular connect:
+    return ParseNetworkAddress ( strAddress, HostAddress, bEnableIPv6 );
+}
+
```
Comment (pljones) [inline src/util.cpp:982]: Could you add a new comment to say something like this expects to find the address either from a valid IP numeric value or a name via DNS lookup through A or AAAA records.

Comment (pljones) [reply src/util.cpp:982]: Oh - note that, given the RFC says the SRV entry can point to an AAAA record, `ParseNetworkAddress` might not quite work.  I think it expects IPv6 addresses to be numeric? `QRegularExpression rx1 ( "^\\[([^]]*)\\](?::(\\d+))?$" );`

File: src/util.cpp (hunk 2 above)
Comment (hoffie) [reply src/util.cpp]: Thansk, updated.

File: src/util.cpp (hunk 3 above)
Comment (hoffie) [reply src/util.cpp]: Ok, changed.

File: src/util.cpp (hunk 4 above)
Comment (hoffie) [reply src/util.cpp:982]: > Could you add a new comment to say something like this expects to find the address either from a valid IP numeric value or a name via DNS lookup through A or AAAA records.

Added.

> Oh - note that, given the RFC says the SRV entry can point to an AAAA record, `ParseNetworkAddress` might not quite work. I think it expects IPv6 addresses to be numeric? `QRegularExpression rx1 ( "^\\[([^]]*)\\](?::(\\d+))?$" );`

Not sure I understand. I think plain IPs in SRV records aren't supposed to be supported. SRV records return names which can be looked up to A or AAAA records. Shouldn't ParseNetworkAddressString handle that properly already?

**Review by ann0see — COMMENTED** [review]
Could you please #ifndef the code? I think some "raw jamulus" people might not want this feature (although it's small)

**Comment by hoffie** [comment]
> Could you please #ifndef the code? I think some "raw jamulus" people might not want this feature (although it's small)

Done now (`QMAKE_CXXFLAGS+="-DCLIENT_NO_SRV_CONNECT"`). I don't think it warrants its own qmake flag though. I've also skipped #ifdef'ing one constant, which will be optimized away by the compiler anyway.

**Review by pljones — APPROVED** [review]

**Comment by ann0see** [comment]
As soon as the DNS records are propagated (and I didn't make a mistake in the format), I can test this.

**Comment by ann0see** [comment]
Ok. Tested ok.

File: src/util.cpp [hunk 5]
```diff
@@ -889,6 +889,99 @@ QSize CMinimumStackedLayout::sizeHint() const
...
+    if ( HostInfo.error() != QHostInfo::NoError )
+    {
+        // qInfo() << qUtf8Printable ( QString ( "Invalid hostname" ) );
+        return false; // invalid address
+    }
+
+    foreach ( const QHostAddress HostAddr, HostInfo.addresses() )
+    {
+        // qInfo() << qUtf8Printable ( QString ( "Resolved network address to %1 for proto %2" ) .arg ( HostAddr.toString() ) .arg (
```
Comment (ann0see) [inline src/util.cpp:906]: Although unrelated, how do we handle debug output in other places at the moment? Commented code, QDebug()?

File: src/util.cpp [hunk 6]
```diff
@@ -889,6 +889,99 @@ QSize CMinimumStackedLayout::sizeHint() const
...
+    QDnsLookup* dns = new QDnsLookup();
+    dns->setType ( QDnsLookup::SRV );
+    dns->setName ( QString ( "_jamulus._udp.%1" ).arg ( strAddress ) );
+    dns->lookup();
+    // QDnsLookup::lookup() works asynchronously. Therefore, wait for
+    // it to complete here by resuming the main loop here.
+    // This is not nice and blocks the UI, but is similar to what
+    // the regular resolve function does as well.
+    QTime dieTime = QTime::currentTime().addMSecs ( DNS_SRV_RESOLVE_TIMEOUT_MS );
+    while ( QTime::currentTime() < dieTime && !dns->isFinished() )
```
Comment (ann0see) [inline src/util.cpp:940]: Not a fan of busy waiting (even if it's limited). Isn't there any blocking function or something like a signal?

File: src/util.cpp [hunk 7]
```diff
@@ -889,6 +889,99 @@ QSize CMinimumStackedLayout::sizeHint() const
...
+    if ( records.length() != 1 )
+    {
+        return false;
+    }
+    QDnsServiceRecord record = records.first();
+    if ( record.target() == "." || record.target() == "" )
+    {
+        // RFC2782 says that "." indicates that the service is not available.
+        // (Qt strips the trailing dot, which is why we check for empty string
+        // as well)
```
Comment (ann0see) [inline src/util.cpp:955]: Does this mean we might have two dots? (When and why) is this valid? Why don't we check just for the empty string?

File: src/util.cpp [hunk 8]
```diff
@@ -889,6 +889,99 @@ QSize CMinimumStackedLayout::sizeHint() const
...
+        // Therefore, end processing here (= return true), but pass back an
+        // invalid HostAddress to let the connect logic fail properly.
```
Comment (ann0see) [inline src/util.cpp:957]: It doesn't ever fail properly in my opinion as there's no user output if we have an invalid IP/Hostname. But that's out of scope.

File: src/util.cpp (hunk 6 above)
Comment (pljones) [reply src/util.cpp:940]: You're asking for a Qt equivalent of AA/AAAA lookup for SRV names?

File: src/util.cpp (hunk 8 above)
Comment (hoffie) [reply src/util.cpp:957]: Right. I've noticed that as well and chose to implement it as close to the existing code path as possible.

File: src/util.cpp (hunk 5 above)
Comment (hoffie) [reply src/util.cpp:906]: Note: This code is unchanged -- it only got moved around and therefore appears in this PR.

I don't see a clear logging policy in the code base, at least not as clear as other (server) software. We shouldn't have commented out code. Instead, we should make proper use of different debug levels and allow controlling them via command line. But yeah, out of scope here...

File: src/util.cpp (hunk 6 above)
Comment (hoffie) [reply src/util.cpp:940]: > Isn't there any blocking function

The existing code path uses [QHostInfo::fromName](https://doc.qt.io/qt-6/qhostinfo.html#fromName), which is a blocking function. I haven't found any equivalent for SRV lookups.

> or something like a signal?

The `QDnsLookup`, which I've used here, is supposed to be used via signals. However, this requires -- as far as I can see -- an inversion of the control flow as the existing logic relies on blocking. Even when that refactoring is done, new questions come up: How should the UI behave? Currently it blocks. If it continues to stay responsive during lookup, we would need some other kind of indicator, so that's kind of a rabbit hole. Not saying that it shouldn't be done, but rather that it's a larger refactoring which I didn't want to do as part of this PR.

In short: That's the best and least intrusive way I've found. If someone finds a better way, I'm certainly open to it. :)

File: src/util.cpp (hunk 7 above)
Comment (hoffie) [reply src/util.cpp:955]: The code basically covers both what the RFC states plus what I've practically observed during testing on Linux. As I haven't tested all other platforms or different DNS implementations, I'd go with this duplicate check -- it's about the failure path anyway. It might be sufficient to check for the empty string, but it would be some effort to prove, I guess?

File: src/util.cpp (hunk 5 above)
Comment (ann0see) [reply src/util.cpp:906]: We can leave it like this for now and look at it in a later refactoring.

File: src/util.cpp (hunk 8 above)
Comment (ann0see) [reply src/util.cpp:957]: Do we have an issue for that? I think I once opened one...

Comment (hoffie) [reply src/util.cpp:957]: #1945 is about improving the connect window error handling. Do you mean that?

Comment (ann0see) [reply src/util.cpp:957]: Yes.

File: src/util.cpp (hunk 7 above)
Comment (ann0see) [reply src/util.cpp:955]: At least add a comment here that the dot check might not be needed.

Comment (hoffie) [reply src/util.cpp]: Done

**Comment by ann0see** [comment]
I still don't feel comfortable with the dot/empty string workaround. The Qt docs don't seem to give much insight. Why does the dot get removed?

**Comment by hoffie** [comment]
> I still don't feel comfortable with the dot/empty string workaround. The Qt docs don't seem to give much insight. Why does the dot get removed?

Qt [uses the OS' libresolv](https://github.com/mburakov/qt5/blob/93bfa3874c10f6cb5aa376f24363513ba8264117/qtbase/src/network/kernel/qdnslookup_unix.cpp#L298) (usually glibc). glibc [has logic to remove a leading dot, which also covers names consisting only of a dot](https://github.com/bminor/glibc/blob/b92a49359f33a461db080a33940d73f47c756126/resolv/dn_expand.c#L99).

**Comment by ann0see** [comment]
Hmm ok. Approving and merging then.

**Review by ann0see — APPROVED** [review]

**Comment by ann0see** [comment]
Dropped needs documentation as the next-release page is up to date: https://github.com/jamulussoftware/jamuluswebsite/blob/next-release/wiki/en/Unregistered-Servers.md#dns-srv-record-support
